ai_edge_torch/convert/conversion_utils.py

```diff
@@ -79,28 +79,30 @@ def flat_arg_names(self) -> list[str]:
     for i in range(args_spec.num_leaves):
       names.append(f"args_{i}")
 
-    dict_context = (
-        kwargs_spec.context
-        if kwargs_spec.type is not collections.defaultdict
-        # ignore mismatch of `default_factory` for defaultdict
-        else kwargs_spec.context[1]
+    kwargs_names = self._flat_kwarg_names(
+        kwargs_spec.children_specs, [kwargs_spec.context]
     )
+    names.extend(kwargs_names)
+    return names
+
+  def _flat_kwarg_names(self, specs, context) -> List[str]:
+    if context is None:
+      return [f"{i}" for i in range(len(specs))] if specs else []
 
-    for name, value_spec in zip(dict_context, kwargs_spec.children_specs):
-      if value_spec.num_leaves == 1:
-        names.append(name)
+    flat_names = []
+    for prefix, spec in zip(context[0], specs):
+      leaf_flat_names = self._flat_kwarg_names(spec.children_specs, spec.context)
+      if leaf_flat_names:
+        flat_names.extend([f"{prefix}_{name}" for name in leaf_flat_names])
       else:
-        # value_spec.num_leaves may be greater than 1 when the value is a (nested)
-        # tuple of tensors. We haven't decided how we should support flattenable
-        # tensor containers as inputs.
-        # TODO(b/352584188): Decide the behavior of tensor container as input (flatten or reject)
-        for i in range(value_spec.num_leaves):
-          names.append(f"{name}_{i}")
-    return names
+        flat_names.append(prefix)
+
+    return flat_names
 
   @property
-  def flat_args(self) -> tuple[torch.Tensor]:
-    return tuple(pytree.tree_flatten(self._normalized_sample_args_kwargs)[0])
+  def flat_args(self) -> tuple[Any]:
+    args, kwargs = self._normalized_sample_args_kwargs
+    return tuple([*args, *kwargs.values()])
```

Comment:
Please keep the original impl which returns the flat tensor arg list. The intent is to make `len(flat_args) == len(flat_arg_names)` as a way to debug when there is an error, and implicitly disable internal flattening in torch.export.export as that may have a conflict to this external flattening.

Reply:
The problem is that during conversion, we use sig.flat_args.
```
  exported_programs: torch.export.ExportedProgram = [
      torch.export.export(sig.module, sig.flat_args, dynamic_shapes=sig.dynamic_shapes)
      for sig in signatures
  ]
  ```
However, with nested arguments, the existing logic would give you the wrong number of keyword arguments.

Please note that on authoring side (i.e, pytorch), the structures are maintained and not flattened.

Reply:
> the existing logic would give you the wrong number of keyword arguments.

Could you explain more on this?

> Please note that on authoring side (i.e, pytorch), the structures are maintained and not flattened.

While the flattening would always happen before the conversion. The intent is to make logics here the only source of truth for the flattening rule. The worst case is we deep flatten the kwargs here to get the flat name list and pass the dict values as is to  torch.export.export, but then torch export uses another rule/order to flatten the dict.

Reply:
Let's take this example:
```
    class SampleModel1(torch.nn.Module):

      def forward(self, x: torch.Tensor, y: torch.Tensor, z: TestContainer):
        return x + y + z.data_1 + z.data_2[0] + z.data_2[1]

    args = (torch.randn(10, 10),)
    kwargs = dict(
        y=torch.randn(10, 10),
        z=TestContainer(
            data_1=torch.randn(10, 10),
            data_2=(torch.randn(10, 10), torch.randn(10, 10)),
        ),
    )
```
On pytorch side, the model would expect **three** arguments (x, y, z). However, the following logic:
```
return tuple(pytree.tree_flatten(self._normalized_sample_args_kwargs)[0])
```
would return **five** values (the flattened args). This triggers an error in `torch.export.export`.

Reply:
oh I see, yes you're right on this. Thanks for the explanation.

We may still run into issues with this impl when the model.forward expects kw-only args. But I guess there are no better way to solve this right now.